### DISCUSSION TITLE
Why does compilation fail only if I use `lexy::parse` and not `lexy::parse_as_tree`?

### DISCUSSION THREAD

**Discussion by EmmanuelMess** [foonathan/lexy#162] [Q&A] [answered]
[Here](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIAMykrgAyeAyYAHI%2BAEaYxBKSAJykAA6oCoRODB7evgGp6ZkCoeFRLLHxUsl2mA5ZQgRMxAQ5Pn6BNXUCDU0EJZExcQnVjc2teR2jfWED5UNVAJS2qF7EyOwc5v5hyN5YANQm/m5OCgTEmKxH2CYaAIJbO3uYh8eoKY4CYte3D2bbDF2XgORzcZ2IYWAP3uj0Bz1ebgAbk08IYCNC/gCgSDjojakRiBjfrDsS9QfRVABPAD0TC6DGpKSaSgAdAgUikif8nsCyccKTTRLRaNE6QBrNkcrlY%2BHkzBU6noBS0SWc/w3GHcuG8hEC6lhFJeAjUmj0VXSnk4twCgD68uNFzSzTtxGIJHN6uJWtJuvlNINRup4MhNoDBA9Goe92YbAUTPW%2BzCBDizFohwA7FZ7vsc/sY5g43SXkwzhms3dc5X9l4MkZ9gBJLCCPA0OKvAAi%2BzO6BAIGDRiO5d%2BVdz4K8Dn2PUcyDLJnT7cHxOzI6753HBH27dqeBYYlny5XuaTicEmGAcUXB8Pq57IHen1ToO7vf7UPV%2ByoxDpD8vFZXc4XfwhyvSsxwnZRUGPOdy2vXNG0YRxW2IRMm0QvALyA4dYM3bddzTVQPBIfBmGTX9sJw5Adz3SlCOIYimFIzCQNzADfywkcwI3Kc8GQABFLwxBbdD0AgqDM3Y694ObJCUIQoSMJg/95zY5ic04ydGmTNhBH3P9YOfEBkQhNFQVEwRSA0hieP4wTWxEyDBGufZkW8TAVL068tJSAxkwRAhKRSRhWBeAAVDFyPlbyeMISytIQiAQq7QLKJoZAADUxC8TAFgk2CQGczLMAgAyWFQPFiuSoT0sKhYFlnZTVIzQDgI8ysa0hScdy8Hz6k0vlOwMvEHBIUEp0wbT0U9JiPIA/ZqWpfYS3RGEr3zQsE2AL8WF3ZDoNy3M1vjF4lTTI5OwFXsTvcw91KktC23yi6QCIMVGBtFI3XQdcsl08izismc0AYM5IuQpgjVQfZiG6/rDgAVisOHOz2xrD3Boh9gQS50A7fYTt7EtKIoEAxBSBAmBtLwGCwYgFDQC4yPI3N0chxo6Fx/GSbpvBidJ8mbXwYBCEp6m4jpkg3OmpncwuAhVgYPHlV7PBUPk4gICxph0AstnaAWFkLiUYhyuCbAAA0AE0bWCesQogcwzH%2B6cHdqxnYIAiAcum/aOM0nj9iBkHVA%2BxaIYK1zcaeksbVfUElt7O61Ysp6jSoAAOO1AQwSEfilpTmqXVrRzXCduJnR6/V7F63o%2BjBvoEX7sKd/3A%2BTYOwbD6H6Fx03Let237bMR2/dMIevcU/SR4DgQg5Dlnw6yyPK5QGeS8mtx45AMvc4nqtWO9xr1K3Si8P2CuqSr1BXoYd7PvrhWUaL33oYnetT3PYhlF6M/9ie85DELBcQQt8670kbtLZugNV6g1DhjLu/UfbYU5iwMINYY54GAA/SwitaDK3fhhE4jl1QQE5oLQgChQTXBZAwVANp6BaxDAALziKgT249EE3Snq3GB88XKLzOr/Ze0ckxngIUmHeHCWLKQPk/KsR9cJiC/s0H%2Bf8vzAyZEAggICvpgMftLNSXDoHt1gZDeBHMlYgFoMLUwxwwAcBZHYpyudsA4N7KID4qwiqkIweQyh6p2GoxXJA6ewM25zzDnw2GUcFAx3ODnY4BlY5TV3oefeLU/qGNCTwzuMNzG4JAIFTAYoSEWKsVomxbg7EAFpHHvgWt4oW69rg5UCfotpzj9glPyWqNwb9kwfyUZNFx1hXF3g%2BF03sPTj5UVoIM5pATZGgUybPDuGNIlLwvivUJL916b2mXhCRjU0mF04Ts/YZkNx6KbsssJqzIYAHcECEALEdPJBNubE2iAYBgYprrXIBiElZJioa5IESYBG4LkbiVaVWeeKs5IyQERMkAbA5iEmOOYAAbItM4vYLnK1VkhJx/hOycx6YnIl/i/nS3ngRVAREwgMVhsi1FFRKFmGxZvfFIA6UMpIpLYZJLRlTIUbQZpbsmbzxovSuijLfJIuQeNNlGKOU4oIHihy6qQDSr5Uy4lpKLEipPt8KlMi2my3lsKzARTkXJmICgkiJBkWYAAI4CVoOgzB%2Bt7zIvJYS9C6LsC1TmtSEKAB5dsYbFroBxiyKxKCCDOs8C0xZ0sOkQHhdJANhxsGc1de6z1WCLDCqaAhLGGRmHoAgLy2V/Kc3Fs5mgbaTB637B1bWplrs87u3nJ7al15gncOMbwwqGzKTvJtEwYAm0zx6uOFyzVhzZHHJWrI9SZcbJWLshc8BA6bnZIxo85561Yac0JjzXsXzDC/LNf86cgLbnArMWCiFSNd3kThf6tsCqLGsoIVizeG73XyXskmAlCKA36uFaCClkHTXpP0Ra4gCsyWjRHsSm475M33XRSk/Ofbb2TwBUO8JazR0COiZO6dFxgBzo3rireI9N0gYuUu1J0iEPP3XLFcaCEVHL2rjfWuOiHzvqCWvBsDAXIqzGhNMTRH729ytm4MNEQhAhWwGlO4wRgX5k9mWKGmA5bIZzcPJlcm8AKDzKgDcaQkwWTdbZbNJB9ibSYGTHie5MD0Amg7QcTVJE5hXVGVNBjiNGNI6Y0FQrwWIwXNC0LVYkMocNWhgGzHt2LuSfNDMbhoPHFY%2B%2BOceW82uhGscN%2B0n0CyYQmx7CHsFkZPC1k4dETyNCqeq3HZccGM1aIZhvDQWOMnJXNSAAVGNjhY39jYEEMQAKi7UZjepIg9dXUerdD6vJuR%2B7WuHqecmE9byuZE0vd8sUc1RnnuJuEe5Vjwj9s4c1oF89n1Cs5nah1DEnV5tQFQb1Hw6GWcTah44fWhmNbvS3CLdyF5RKETEqxZwetaqEOtqym250DcQcFyss0ctuebYSGE84Q0njtY%2BVdJJZTHFNAWSkIMWBEk1DKHUcoFTfiyIyZkmBJ0xPONaiMXpWdWj1IiSzwHmFC5Z5aPk1o/R2lUMaJtKQ6BxEV95CWxBpeYll76KkiuHSYCdFouIbptfsjVJGX443fiIkgjjL621KT6auTmeeTICAIFxgZOnCgGdaUmQxBABHd5woYIaS5HXl4XC1jaOn7Px0gFThnRgaBiJvmwBAT3CAWTIFierLtYfO4Fm6lH86y8NFKFBCI5DYheyE52r2NHLBuoY4YGNa4GaI9GhZNELwVAqBxE9snBX9peyOhIKbsr2uc/Z%2BD3ngvbCFkSRbJ0w2Ze2TR0iWw7brdgVO1hhv2g4Yd%2BQ6GwuEnl%2B7jLd%2BMeXcYQ%2B0k%2BLxjMMY7DduNQCwVX9BiAa9oFrqGD3omivleE9FXrzgLrzvwOim4FgLsP5IFN3pHl7C4lARKvPG0lgcfstMutHpshAXzjaOgccLXqmA3ltE3lvOjg%2BJ3sQlARZGGKPgbuPiAJPs6GbiQKAR5E9OLgoJLkVN2CsAQDrBcJgBZNBLwRLoJMwvHoYMgJSABNwZWGvhAGAGADgSmoeMlvsFwGxPOBwEsLghwHDLwH4BwFoKQKgJwG4NYCMgoCsGsGSP8DwKQOqhYYYUsGKCAHDBoPoJwJIGYZoLwNYRwLwAoCAH4e4VoEsHALADAIgCvN/mrmQBQBACrikSAMAFwP8KQFgOLusGlOhPcmGoFOYa4aaHahERANEMEaQNEIyvNpwK4Q0cwPNmGtENoPiHUU2hNGGgwLQOOh4XkZgP3sABvMKBEdwLwFgLuEYOIMMfgBcHUHiFMZYfKLUEaBsJYSIsYa4VYtEF%2BPNh4FgHUXEiwM0YYXwAYMAAoEUZgCUWUZcTIIICIGIOwFIC8fIEoGoHUboFwPoIYMYHYZYPoHgNEBEbAPmJYn6KQHiPEGuD8rVFYR8FkFMVUt2GdKPFYJYGYJYWVGbvCpCSPp0A%2BC4NTOMH4ACSEDMGUBUHoGkBkA%2BJSQyYUA%2BP0HSfMLYN5viL1L0CyQCaSXyc0ByYMJULYFMAKRKb0KKWihIEsA4asOsPKf4SYUEcMaEfsKoGnJilUpipIK5sgDODkSyP4J0rgIQC5lsFwAsLwNEZ4aQN4b4aqYEaQBcc6eYfiZwOEZEW4cEbEQkRrCABscgBDKkZQE0LcV/OELQEIAgKgPcuUbwBkRtgwNGd5nGQmZ6cmV/j/kMLccwCkAoPGSISmXEBEMFN6aQGWcQGGkaJmYmXUSGXcMQLcVWSGQ0MRJwMEuWMEncLwPwK8UKB8dIIOd8SoOoMMf8YCUYCgCCTYAccSRANCQKHCUMIiWKMifeGiZwBiQQOgFifOXiSEfCRCFgMSQqY4cqYKfuWEOmbGfGY2cMVAb6fcl%2BCkJcUYZwKYaQNmVYZwNgKoJsQSFqTqXqQacAEaZ0huXVBALYZYNYBZBaSBdabaX6R4cicYa6e6X4X%2BaET6VEf6V4T4X4cYf4OqV6WEehTEV%2BRwGYBRSEVWfacifCRkM4JIEAA%3D%3D%3D) is a working example of the error. If I uncomment dummy() the program no longer compiles, but I am not sure what broke.

**Comment by foonathan** [comment, marked as answer]
You've only used `lexy::parse_as_tree` before, which completely ignores all `value` callbacks. With `lexy::parse` reading from a file, the program no longer compiles because you are now instantiating the callbacks.

Your issue is that the `Static` production produces a value of the `Static` tag type, but nobody uses it. You can either use `lexy::noop` as the value to not produce a tag type and drop it entirely (https://godbolt.org/z/nPh1dM9aq), store the tag type in the datastructures, or just inline the rule from `Static` anywhere its used.

**Reply by EmmanuelMess** [reply]
The error needs to be more clear IMO :sweat_smile:

**Reply by foonathan** [reply]
If we skip the template callstack, the error message is:

```
/opt/compiler-explorer/libs/lexy/trunk/include/lexy/grammar.hpp:355:36: error: static assertion failed: missing value callback overload for production
  355 |             static_assert(_detail::error<Production, Args...>,
      |                           ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/compiler-explorer/libs/lexy/trunk/include/lexy/grammar.hpp:355:36: note: 'lexy::_detail::error<internal::grammar::StaticQualifiedPoint, internal::ast::Static, lexy::member<lexy::_mem_ptr_fn<&internal::ast::StaticQualifiedPoint::identifier> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >' evaluates to false
```

This tells you that the callback of `StaticQualifiedPoint` does not accept the values `ast::Static, lexy::member<...>, std::string`. What information would you like to have?

**Reply by EmmanuelMess** [reply]
The readable part is:

> error: static assertion failed: missing value callback overload for production

Which has no info about which callback, which overload or which production (the note doesn't help with this because there is no clear indication of what check actually evaluated to false). 

A better error would be:

> error: static assertion failed: internal::grammar::StaticQualifiedPoint::value is missing a value callback overload for production internal::grammar::Static: valued productions that are captured in a rule cannot be ignored in value

**Reply by foonathan** [reply]
That's not possible before C++26, unfortunately. You just have to learn how to extract the relevant information from the error message and the callback and rule documentation, sorry.
